docs/site/migration/models/remoting-hooks.md

```diff
@@ -6,8 +6,523 @@ sidebar: lb4_sidebar
+
+  next();
+});
+```
+
+This remoting hook can be rewritten to a LoopBack 4 interceptor as follows:
+
+```ts
+async function intercept(
+  invocationCtx: InvocationContext,
+  next: () => ValueOrPromise<InvocationResult>,
+) {
+  const result = await next();
+  if (Array.isArray(result)) {
+    result.forEach(it => delete it.password);
+  } else {
+    delete result.password;
+  }
+  return result;
+}
+```
+
+## Rejecting requests
+
+In order to reject a request and return an error HTTP response, just throw an
+error from your interceptor. As explained in
+[Handling errors in controllers](../../Controllers.html#handling-errors-in-controllers),
+you can use one of `HttpErrors` constructors to control the HTTP status code of
+the response.
+
+Consider the following LoopBack 3 example rejecting requests from non-local
+addresses:
+
+```js
+Customer.beforeRemote('**', function(ctx, next) {
+  if (ctx.req.remoteAddress !== '127.0.0.1') {
+    next(new HttpErrors.Forbidden());
+  } else {
+    next();
+  }
+});
+```
+
+This remoting hook can be rewritten to a LoopBack 4 interceptor as follows:
+
+```ts
+class AddRemoteAddressInterceptor implements Provider<Interceptor> {
+  constructor(@inject(RestBindings.Http.REQUEST) private request: Request) {}
+
+  // ...
+
+  async intercept(
+    invocationCtx: InvocationContext,
+    next: () => ValueOrPromise<InvocationResult>,
+  ) {
+    if (this.request.socket.remoteAddress !== '127.0.0.1') {
+      throw new HttpErrors.Forbidden();
+    }
+    return next();
+  }
+}
+```
+
+## Accessing the current user
+
+LoopBack 4 does not provide authentication layer out of the box, it relies on
+extensions to implement authentication features. Please refer to the
+documentation for your authentication component to learn how to access data
+about the current user.
+
+If you are using the official
+[LoopBack Authentication Extension](../../lb4/Loopback-component-authentication.html),
+then you can access profile of the currently authenticated user via the binding
+`SecurityBindings.USER`.
+
+Please note the value for the binding may be bound later in the request handling
+cycle, therefore it's best to use `@inject.getter()` to defer resolution of the
+value until the intercept method is invoked.
+
+Consider the following LoopBack 3 example printing a warning whenever an
+anonymous (unauthenticated) request is made:
+
+```js
+Customer.beforeRemote('*.save', function(ctx, unused, next) {
+  if (!ctx.req.accessToken) {
+    console.warn('anonymous request!');
+  }
+  next();
+});
+```
+
+This remoting hook can be rewritten to a LoopBack 4 interceptor as follows:
+
+```ts
+class WarnAnonymousInterceptor implements Provider<Interceptor> {
+  constructor(
+    @inject.getter(SecurityBindings.USER)
+    private getCurrentUser: Getter<UserProfile>,
+  ) {}
+
+  async intercept(
+    invocationCtx: InvocationContext,
+    next: () => ValueOrPromise<InvocationResult>,
+  ) {
+    const currentUser = await this.getCurrentUser();
```

Comment:
I tried to test this example in loopback4-shopping-example by creating a global interceptor, but received the following error when trying to invoke the `ping` method:

```
Unhandled error in GET /ping: 500 Error: The key 'security.user' is not bound to any value in context application
    at ShoppingApplication.getBinding (packages/shopping/node_modules/@loopback/context/dist/context.js:604:15)
    at RestServer.getBinding (packages/shopping/node_modules/@loopback/context/dist/context.js:600:33)
    at RequestContext.getBinding (packages/shopping/node_modules/@loopback/context/dist/context.js:600:33)
    at InterceptedInvocationContext.getBinding (packages/shopping/node_modules/@loopback/context/dist/context.js:600:33)
    at InterceptedInvocationContext.getValueOrPromise (packages/shopping/node_modules/@loopback/context/dist/context.js:659:30)
    at InterceptedInvocationContext.get (packages/shopping/node_modules/@loopback/context/dist/context.js:581:21)
    at WarnAnonymousInterceptor.getter [as getCurrentUser] (packages/shopping/node_modules/@loopback/context/dist/inject.js:258:20)
    at WarnAnonymousInterceptor.intercept shopping/dist/interceptors/warn-anonymous.interceptor.js:40:40)
    at packages/shopping/node_modules/@loopback/context/dist/interceptor-chain.js:101:20
```

I thought our shopping example app is already configured with the authentication extension and therefore `SecurityBindings.USER` should be available to request interceptors. What am I missing?

Here is the full source code of the interceptor, you can save it to `packages/shopping/src/interceptors/warn-anonymous.interceptor.ts` to reproduce the issue yourselves.

```ts
import {
  globalInterceptor,
  Interceptor,
  InvocationContext,
  InvocationResult,
  Provider,
  ValueOrPromise,
  inject,
  Getter,
} from '@loopback/context';
import {SecurityBindings, UserProfile} from '@loopback/security';

/**
 * This class will be bound to the application as an `Interceptor` during
 * `boot`
 */
@globalInterceptor('', {tags: {name: 'WarnAnonymous'}})
export class WarnAnonymousInterceptor implements Provider<Interceptor> {
  constructor(
    @inject.getter(SecurityBindings.USER)
    private getCurrentUser: Getter<UserProfile>,
  ) {}

  /**
   * This method is used by LoopBack context to produce an interceptor function
   * for the binding.
   *
   * @returns An interceptor function
   */
  value() {
    return this.intercept.bind(this);
  }

  /**
   * The logic to intercept an invocation
   * @param invocationCtx - Invocation context
   * @param next - A function to invoke next interceptor or the target method
   */
  async intercept(
    invocationCtx: InvocationContext,
    next: () => ValueOrPromise<InvocationResult>,
  ) {
    const currentUser = await this.getCurrentUser();
    if (!currentUser) {
      console.warn('anonymous request!');
    }
    return next();
  }
}
```

Reply:
Do you see this error when a method with `@authenticate` is invoked from REST? I think we only bind SecurityBindings.CURRENT_USER for routes that require authentication.

Reply:
> I thought our shopping example app is already configured with the authentication extension and therefore SecurityBindings.USER should be available to request interceptors. What am I missing?

@bajtos This could be caused by no user profile is set in the [authentication action](https://github.com/strongloop/loopback-next/blob/master/packages/authentication/src/providers/auth-action.provider.ts#L64)

And many reasons could result in an undefined user profile. For example: If you visit an endpoint without including a token in the request header, OR your endpoint is not decorated with `@authenticate()`, OR the jwt service cannot decode a valid user profile from the token(an invalid or expired token).

BTW, when is the interceptor being executed? After the auth action and before invoking the controller method?

Reply:
> BTW, when is the interceptor being executed? After the auth action and before invoking the controller method?

Yes.

Reply:
Thank you all for the comments.

> And many reasons could result in an undefined user profile. For example: If you visit an endpoint without including a token in the request header, OR your endpoint is not decorated with @authenticate(), 

In these cases, I would expect `SecurityBindings.CURRENT_USER` to be bound to `undefined` or `null` value, so that consumers of this binding (e.g. my example interceptor) can distinguish between anonymous and authenticated requests.

>the jwt service cannot decode a valid user profile from the token(an invalid or expired token).

In this case, I would expect an error response telling the HTTP client about the problem with the token.

Reply:
The doc issue discussed in this thread should have been resolved by 49f5e8c08.